src/transformers/models/groupvit/modeling_groupvit.py

```diff
@@ -0,0 +1,1864 @@
+        if key is None:
+            key = query
+        S = key.size(1)
+
+        # [B, nh, N, C//nh]
+        q = self.q_proj(query).reshape(B, N, self.num_heads, C // self.num_heads).permute(0, 2, 1, 3)
+        # [B, nh, S, C//nh]
+        k = self.k_proj(key).reshape(B, S, self.num_heads, C // self.num_heads).permute(0, 2, 1, 3)
+        # [B, nh, S, C//nh]
+        v = self.v_proj(key).reshape(B, S, self.num_heads, C // self.num_heads).permute(0, 2, 1, 3)
+
+        # [B, nh, N, S]
+        attn = (q @ k.transpose(-2, -1)) * self.scale
+        attn = attn.softmax(dim=-1)
+        assert attn.shape == (B, self.num_heads, N, S), f"{attn.shape} vs {(B, self.num_heads, N, S)}"
+
+        # [B, nh, N, C//nh] -> [B, N, C]
+        out = (attn @ v).transpose(1, 2).reshape(B, N, C)
+        out = self.proj(out)
+        return out
+
+
+class CrossAttnBlock(nn.Module):
+    def __init__(
+        self,
+        config: GroupViTVisionConfig,
+    ):
+        super().__init__()
+        self.attn = Attention(config)
+        self.norm2 = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.mlp = GroupViTMLP(config)
+        self.norm_post = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+
+    def forward(self, query, key):
+        x = query
+        x = x + self.attn(query, key)
+        x = x + self.mlp(self.norm2(x))
+        x = self.norm_post(x)
+        return x
+
+
+class GroupViTAssignAttention(nn.Module):
+    def __init__(
+        self,
+        config: GroupViTVisionConfig,
+    ):
+        super().__init__()
+        self.scale = config.hidden_size**-0.5
+
+        self.q_proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.k_proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.v_proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.assign_eps = config.assign_eps
+
+    def get_attn(self, attn, gumbel=True, hard=True):
+
+        if gumbel and self.training:
+            attn = gumbel_softmax(attn, dim=-2, hard=hard)
+        else:
+            if hard:
+                attn = hard_softmax(attn, dim=-2)
+            else:
+                attn = F.softmax(attn, dim=-2)
+
+        return attn
+
+    def forward(self, query, key):
+        B, N, C = query.shape
+        value = key
+        S = key.size(1)
+        # [B, N, C]
+        q = self.q_proj(query)
+
+        # [B, S, C]
+        k = self.k_proj(key)
+
+        # [B, S, C]
+        v = self.v_proj(value)
+
+        # [B, N, S]
+        raw_attn = (q @ k.transpose(-2, -1)) * self.scale
+
+        attn = self.get_attn(raw_attn)
+        soft_attn = self.get_attn(raw_attn, gumbel=False, hard=False)
+
+        attn = attn / (attn.sum(dim=-1, keepdim=True) + self.assign_eps)
+
+        assert attn.shape == (B, N, S)
+
+        out = attn @ v
+
+        out = self.proj(out)
+
+        return out, soft_attn
```

Comment:
Here, also it would be great if you can replace this like B -> batch_size, C -> num_channels etc., to make the code more readable